### ISSUE TITLE
perf(slh-dsa): make the verify compile cost payable once, not once per process

### ISSUE THREAD

**Issue by chokobole** [fractalyze/sig-frx#82]
### Pre-submission checklist

- [x] I have searched the existing issues and this is not a duplicate.

### Context / problem

`frx.jit(scheme.verify)` compiles once per `(parameter set, batch size)`: **55 s
for SLH-DSA-SHA2-128f and 14 s for 128s** on CPU, measured by `verify_bench`.
Warm, the traced path beats eager at every batch size measured — 8.8–13.1× at
B=1, 1.11–1.18× at B=64 — so the compile is the only thing a caller pays that
eager verification does not.

#14 closed on the judgment that this, not a second host-only verifier, is the
entry worth defending. The only caller the eager path still serves better is one
that cannot amortize 55 s, and the answer for them is a cache rather than a
second implementation that has to agree byte for byte.

Nothing in this repo tells that caller how. `verify_bench` reports compile
separately from warm latency — which is what surfaced the number — but neither it
nor the scheme page says the cost is avoidable across processes, or whether frx
even honors the knob that would avoid it.

### Proposed solution

Measure, then document. No code in the hot path.

1. Set `JAX_COMPILATION_CACHE_DIR` and run `verify_bench` twice in **separate
   processes**, both category 1 sets, recording compile cold versus warm. An
   in-process repeat does not answer the question — the point is a caller who
   restarts.
2. If frx honors it, document it where a caller looks: the **"Where the batch
   axis is"** section of `docs/schemes/slh-dsa.md`, and `verify_bench`'s
   docstring, which already explains that a compile is paid once per shape.
3. If frx does not honor it, that is the finding and this becomes an frx issue
   rather than a sig-frx one. Record the number either way.

### Alternatives considered

- **A `--compile_cache_dir` flag on `verify_bench`, or a sig-frx config knob.**
  Rejected on evidence rather than taste: the same wrapper was built and closed in
  a sibling repo (`fractalyze/openvm-zorch` PR #74, closed as redundant) because
  it re-wraps a generic JAX knob once per consumer. JAX reads the environment
  variable natively, so the deliverable here is a measurement plus a sentence in
  the docs. If frx does not honor it, the fix belongs in frx.
- **Keep a host-only verification path instead.** The alternative #14 rejected,
  recorded there in full: it buys 28.5% (128f) / 35.0% (128s) at B=64 for a
  caller who never compiles, nothing for one who does, and costs two verifiers
  that must agree byte for byte.
- **Do nothing.** Defensible — a service verifying signatures compiles once at
  start-up and never notices. It leaves the one caller #14's decision knowingly
  under-serves with no answer at all, which is worth the measurement it takes to
  close.

### Why this approach

A decision is only as good as its remedy. #14 declined a second verifier on the
grounds that compile time is the thing to defend; leaving that undefended and
undocumented makes the decision look like a preference rather than a trade with a
known other side.

Measuring first is also what says whether there is any work here at all. If frx
honors the variable, this is two numbers and a paragraph. If it does not, the
issue moves rather than growing — and either outcome is cheaper than the second
verifier it stands in for.

### Plan

- [x] Measure `verify_bench` compile with and without a persistent cache dir, in separate processes, both category 1 sets
- [x] Record the numbers here whichever way they go
- [x] If honored: document it on `docs/schemes/slh-dsa.md` and in `verify_bench`'s docstring
- [ ] If not honored: file the frx-side issue and close this one pointing at it

### Acceptance criteria

- [x] A second process compiling the same `(parameter set, batch size)` is measured, and the number is on this issue
- [ ] A caller can find out how to avoid the compile without reading this issue
- [x] No sig-frx-local re-wrap of a generic frx / JAX knob

### References

- The decision this defends, with the latency and compile numbers: #14
- Sibling-repo precedent for rejecting a repo-local flag: `fractalyze/openvm-zorch` PR #74 (closed as redundant)

### Area

area:slh-dsa

### Parent issue

#14

### Milestone

Hash-based signatures

### Relevant knowledge

Two items from personal notes, inlined in full rather than linked — neither is in
the team wiki.

**A persistent compilation cache is a JAX-native environment variable, and
wrapping it per repo was rejected once already.** JAX reads
`JAX_COMPILATION_CACHE_DIR` directly, equivalently
`jax.config.update("jax_compilation_cache_dir", …)`. Measured in
`fractalyze/openvm-zorch` #70 on a GPU prover: a warm cache took one stage's cold
compile from 418 s to 95 s and the whole prove from 844 s to 295 s, with no code
change. The `--compile_cache_dir` flag proposed there (openvm-zorch PR #74) was
closed as a redundant per-repo re-wrap of a generic knob. Also measured there and
relevant to the "if not honored" branch: a second call **in the same process**
re-compiled the large kernel rather than hitting XLA's in-process executable
cache, so a persistent cache can matter even within one run.

**Tracing wins most where the batch is small, which is why the compile is worth
defending.** `verify_bench`, traced against eager-on-main, one workstation:

| set | B | eager (main) | traced | |
|---|---|---|---|---|
| SHA2-128f | 1 | 288.1 ms | 32.7 ms | 8.8× |
| | 16 | 425.0 ms | 138.6 ms | 3.1× |
| | 64 | 859.9 ms | 776.0 ms | 1.11× |
| SHA2-128s | 1 | 135.8 ms | 10.4 ms | 13.1× |
| | 16 | 215.3 ms | 49.3 ms | 4.4× |
| | 64 | 318.8 ms | 269.8 ms | 1.18× |

The eager dispatch count is a function of the parameter set alone and never of
`B`, so eager is dispatch-bound at small `B` (41% host at 128f B=1) and hash-bound
at large `B` (80% hash at B=64). Tracing removes exactly the fixed cost a large
batch was already amortizing. The caller that gains an order of magnitude is the
one verifying a signature at a time — and that caller is also the least able to
amortize a 55 s compile, which is the whole of this issue.

**Comment by chokobole** [comment]
### Session 2026-08-06 16:30 progress

Measured. **FRX honors `JAX_COMPILATION_CACHE_DIR` across processes**, so this is the "two numbers and a paragraph" branch rather than the frx-issue branch.

`verify_bench --batches=1,64`, both category 1 sets, three separate processes — control (no cache dir), cold (fresh dir), warm (same dir, new process). One CPU workstation, load ~1.5, `JAX_RAISE_PERSISTENT_CACHE_ERRORS=1` throughout so a cache failure would raise rather than read as a miss.

| set | B | control | cold | warm | |
| --- | --- | --- | --- | --- | --- |
| SHA2-128f | 1 | 105.6 s | 109.2 s | **21.9 s** | −79.9% |
| SHA2-128f | 64 | 94.1 s | 92.4 s | **22.2 s** | −76.0% |
| SHA2-128s | 1 | 35.0 s | 34.3 s | **4.4 s** | −87.2% |
| SHA2-128s | 64 | 35.1 s | 34.4 s | **4.7 s** | −86.3% |

Three readings worth keeping:

- **Warm latency does not move** — 0.48 against 0.51 ms/sig at 128f B=64, 0.16 either way at 128s B=64. The cache displaces compilation and nothing else, so it costs a caller nothing at runtime.
- **The cache held exactly four entries, 16 MB**, named `jit_verify-<hash>-cache`. That is this issue's `(parameter set, batch size)` claim made visible on disk: two sets at two batch sizes.
- **The residual is not zero.** ~22 s and ~4.5 s remain because the cache is keyed on the lowered HLO — tracing and lowering happen in every process regardless. The knob removes the XLA compile, not the whole cold start. Worth knowing before quoting it as "the compile goes away".

Compile here is ~94 s at 128f against the 55 s this issue quotes — different box or load. Reported as measured rather than reconciled.

#### The measurement raised a correctness question, so it got gated

A CPU cache hit logs, on every executable load:

```
E cpu_aot_loader.cc:195] Loading XLA:CPU AOT result. Target machine feature
+prefer-no-gather is not supported on the host machine. ... This could lead to
execution errors such as SIGILL.
```

It fires on the machine that *wrote* the entry. `+prefer-no-gather` is a code-generation preference that appears in the compile's feature string and never in host detection, so the two cannot match. But `verify_bench` asserts no verdicts, so nothing in the table above shows that a cache-loaded executable still verifies correctly — and a knob documented as "set this to skip the compile" that silently returned wrong verdicts would be a worse outcome than leaving this undocumented. Gated separately on the traced path's own vectors:

```
slh_dsa_traced_test, cache cold:  PASSED in 166.5 s
slh_dsa_traced_test, cache warm:  PASSED in  38.7 s   (−76.8%)
```

Entry count unchanged across the warm run, so those were pure hits. The vectors pass, so on one host the warning is noise — and across unlike hosts it is the reason a cache directory should not be shared between them. Both are in the doc.

#### A second finding, which is most of why this was undiscoverable

**The fork keeps JAX's spelling.** The knob is `optional_string_state(name='jax_compilation_cache_dir')` and the environment variable is that name uppercased, so FRX reads `JAX_COMPILATION_CACHE_DIR`. The `FRX_COMPILATION_CACHE_DIR` a caller would reasonably guess sets nothing and reports nothing. Documented explicitly.

#### Done

- PR #97 — `docs/schemes/slh-dsa.md` gets the numbers, the naming trap, the residual and the warning; `verify_bench`'s docstring gets the knob and how to re-measure; `docs/README.md` index row updated so the hub does not drift.
- No `--compile_cache_dir` flag, per this issue and the openvm-zorch PR #74 precedent.

Two boxes deliberately left open: *"If not honored: file the frx-side issue"* is the branch that did not fire, and *"A caller can find out how to avoid the compile without reading this issue"* ticks when #97 merges.

**One judgment call for review.** This issue asked for the docs in the "Where the batch axis is" section; #97 gives it its own `## What the trace costs, and how to pay it once` instead. The content runs ~40 lines and is about compile economics rather than where the batch axis is, and a caller scanning headings for how to avoid the compile finds it there. `conventions.md` specifies three questions "and everything else is optional", so a fourth section is permitted rather than a skeleton deviation — but it is a departure from this issue's letter, so it is easy to fold in if the stricter reading is preferred.

- Next session: review and merge #97; nothing else outstanding here.
- Blocker: none.

**Comment by chokobole** [comment]
### Correction — `FRX_COMPILATION_CACHE_DIR` works; the "naming trap" above is wrong

The previous comment claimed the fork keeps JAX's spelling and that
`FRX_COMPILATION_CACHE_DIR` "sets nothing and reports nothing". That is false, and
it was the one part of that comment I had not measured.

`frx/__init__.py` mirrors the prefix before the config layer reads the environment:

```python
for _k, _v in list(_os.environ.items()):
    if _k.startswith("FRX_"):
        _os.environ.setdefault("JAX_" + _k[4:], _v)
```

Prefix-wide rather than an enumerated list, so it covers every `JAX_*` config
variable including ones upstream adds later. Setting each prefix alone, nothing
else, writes the identical cache entry to disk:

```
FRX_COMPILATION_CACHE_DIR only -> jit__lambda-71a00d45...aac8-cache
JAX_COMPILATION_CACHE_DIR only -> jit__lambda-71a00d45...aac8-cache
```

`JAX_*` wins when both are set, but only because the mirror uses `setdefault`.

**How the error happened**, since it is the reusable part: I read the env var's
derivation in `_src/config.py` — `os.getenv(name.upper())` over the state name
`jax_compilation_cache_dir` — and concluded the `FRX_` spelling was dead, without
checking the package `__init__.py` above it. The measurement itself never caught
it because I ran every pass under `JAX_*` and never tried the other spelling. A
two-line A/B would have settled it before it reached a doc.

**What this changes:** the documented spelling, nothing else. The compile numbers,
the four-entry cache, the residual, the `SIGILL` warning and the traced-KAT gate
all stand — one variable, two names, and the runs used one of them.

PR #97 now names `FRX_COMPILATION_CACHE_DIR` as the spelling to use, with a
sentence on the mirror so a reader arriving from JAX's own docs knows the two are
the same knob.

**Comment by chokobole** [comment]
### Session 2026-08-06 20:40 — closed by #97 (8598f08)

- **Done:** measured, then documented. FRX honors a persistent compilation cache across processes; compile drops ~4–8× (numbers, method and the traced-KAT correctness gate are in the two comments above). Landed as **10 lines**, not the 54 the first revision had.
- **Next session:** nothing on this issue. See the follow-up below, which is not yet filed anywhere.
- **Blocker:** none.

#### Why the change shrank from 54 lines to 10

`docs/reference/conventions.md` opens by refusing most of what the first revision added:

> The rules every FRX consumer shares … are not repeated here. They are identical in every repo built on FRX, and **a copy per repo is exactly how they drift apart.**

`hash-frx` already documents this knob under *"The compile cache is per toolchain"*, including a `--test_env` caveat sig-frx never had. And the two copies had **already drifted before this landed**: hash-frx scopes the don't-share rule to one cache directory per frxlib build; the draft here scoped it to unlike hosts. Both true, neither complete — the predicted failure, reproduced inside one PR.

Two items were removed as **wrong**, not merely misplaced:

- *"two parameter sets at two batch sizes leave four entries"* described one `--batches=1,64` invocation as design. `_BATCHES` defaults to seven values, so the default run leaves fourteen.
- The `+prefer-no-gather` / `SIGILL` paragraph told readers to ignore a warning that becomes meaningful the moment upstream fixes the feature-string/host-detection mismatch — against an XLA fork this team edits, with no link from that code back to the sentence.

What shipped is the part a scheme page owes a reader: parameter set is *this scheme's* axis, the taller hypertrees are why the compile is tens of seconds, and the caller who gains most from tracing amortizes it over the fewest calls.

#### Follow-up, unfiled — the knob has no canonical home

`FRX_COMPILATION_CACHE_DIR` guidance now sits in two consumer repos (`hash-frx/docs/reference/development.md` and, one sentence of it, `sig-frx/docs/schemes/slh-dsa.md`), and neither is FRX-consumer-wide. The drift above is the cost, and it will recur in `enc-frx` / `zorch` / `openvm-zorch` the next time someone needs the knob. Worth one page in FRX itself that all consumers point at, covering: the variable and that `FRX_*` mirrors onto `JAX_*` prefix-wide (`setdefault`, so `JAX_*` wins); one cache directory per frxlib build; the `--test_env` gap under Bazel; and that the persistent cache is keyed on lowered HLO, so tracing and lowering are re-paid every process.

Not filed — it belongs in the FRX repo, and that is a decision for whoever owns those docs.
